File: android-sdk/sdk/src/main/java/com/adlcom/mxl/sdk/compliance/GdprManager.kt
package com.adlcom.mxl.sdk.compliance

import android.content.Context
import com.adlcom.mxl.sdk.internal.SdkState
import com.adlcom.mxl.sdk.internal.getSdkState
import com.adlcom.mxl.sdk.logging.Logger
import com.adlcom.mxl.sdk.storage.StorageManager
import com.google.gson.Gson
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/**
 * GDPR compliance manager for data privacy.
 */
object GdprManager {
    private const val PREFS_NAME = "mxl_gdpr_prefs"
    private const val KEY_CONSENT_GIVEN = "consent_given"
    private const val KEY_CONSENT_TIMESTAMP = "consent_timestamp"

    /**
     * Check if user has given consent.
     */
    fun hasConsent(context: Context): Boolean {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        return prefs.getBoolean(KEY_CONSENT_GIVEN, false)
    }

    /**
     * Set user consent.
     */
    fun setConsent(context: Context, hasConsent: Boolean) {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        prefs.edit()
            .putBoolean(KEY_CONSENT_GIVEN, hasConsent)
            .putLong(KEY_CONSENT_TIMESTAMP, System.currentTimeMillis())
            .apply()
        
        if (!hasConsent) {
            // Stop data collection
            Logger.d("User consent withdrawn, stopping data collection")
        }
    }

    /**
     * Export user data (GDPR right to data portability).
     */
    fun exportUserData(userId: String, callback: (String?) -> Unit) {
        val scope = CoroutineScope(Dispatchers.IO)
        scope.launch {
            try {
                val sdkState = getSdkState() ?: run {
                    callback(null)
                    return@launch
                }

                // Collect all user data
                val userData = mutableMapOf<String, Any>()
                
                // Session data
                val sessionId = sdkState.sessionManager.getSessionId()
                if (sessionId != null) {
                    userData["sessionId"] = sessionId
                }
                
                // User attributes
                val userId = sdkState.sessionManager.getUserId()
                if (userId != null) {
                    userData["userId"] = userId
                }
                
                // TODO: Export events from database
                // TODO: Export crash reports
                // TODO: Export performance data
                
                val json = Gson().toJson(userData)
                callback(json)
            } catch (e: Exception) {
                Logger.e("Error exporting user data", e)
                callback(null)
            }
        }
    }

    /**
     * Delete user data (GDPR right to be forgotten).
     */
    fun deleteUserData(userId: String, callback: (Boolean) -> Unit) {
        val scope = CoroutineScope(Dispatchers.IO)
        scope.launch {
            try {
                val sdkState = getSdkState() ?: run {
                    callback(false)
                    return@launch
                }

                // TODO: Delete all user events from database
                // TODO: Delete crash reports
                // TODO: Delete performance data
                // TODO: Clear session data
                
                // Clear user identification
                sdkState.sessionManager.identify("", emptyMap())
                
                Logger.d("User data deleted for user: $userId")
                callback(true)
            } catch (e: Exception) {
                Logger.e("Error deleting user data", e)
                callback(false)
            }
        }
    }

    /**
     * Anonymize user data.
     */
    fun anonymizeUserData(userId: String, callback: (Boolean) -> Unit) {
        val scope = CoroutineScope(Dispatchers.IO)
        scope.launch {
            try {
                val sdkState = getSdkState() ?: run {
                    callback(false)
                    return@launch
                }

                // TODO: Anonymize all user events
                // Replace userId with anonymous ID
                val anonymousId = "anonymous_${System.currentTimeMillis()}"
                
                Logger.d("User data anonymized for user: $userId")
                callback(true)
            } catch (e: Exception) {
                Logger.e("Error anonymizing user data", e)
                callback(false)
            }
        }
    }
}

File: ios-sdk/Sources/MxLSdk/Compliance/GdprManager.swift
import Foundation

/**
 * GDPR compliance manager for data privacy.
 */
internal class GdprManager {
    private enum Keys {
        static let consentGiven = "mxl_consent_given"
        static let consentTimestamp = "mxl_consent_timestamp"
    }
    
    /**
     * Check if user has given consent.
     */
    static func hasConsent() -> Bool {
        return UserDefaults.standard.bool(forKey: Keys.consentGiven)
    }
    
    /**
     * Set user consent.
     */
    static func setConsent(_ hasConsent: Bool) {
        UserDefaults.standard.set(hasConsent, forKey: Keys.consentGiven)
        UserDefaults.standard.set(Date().timeIntervalSince1970, forKey: Keys.consentTimestamp)
        
        if !hasConsent {
            Logger.shared.info("User consent withdrawn, stopping data collection")
        }
    }
    
    /**
     * Export user data (GDPR right to data portability).
     */
    static func exportUserData(userId: String) async -> String? {
        do {
            let sdkState = MxLSdk.getSdkState()
            guard let sdkState = sdkState else { return nil }
            
            var userData: [String: Any] = [:]
            
            // Session data
            if let sessionId = sdkState.sessionManager.getSessionId() {
                userData["sessionId"] = sessionId
            }
            
            // User attributes
            if let userId = sdkState.sessionManager.getUserId() {
                userData["userId"] = userId
            }
            
            // TODO: Export events from database
            // TODO: Export crash reports
            // TODO: Export performance data
            
            let jsonData = try JSONSerialization.data(withJSONObject: userData, options: .prettyPrinted)
            return String(data: jsonData, encoding: .utf8)
        } catch {
            Logger.shared.error("Error exporting user data", error: error)
            return nil
        }
    }
    
    /**
     * Delete user data (GDPR right to be forgotten).
     */
    static func deleteUserData(userId: String) async -> Bool {
        do {
            let sdkState = MxLSdk.getSdkState()
            guard let sdkState = sdkState else { return false }
            
            // TODO: Delete all user events from database
            // TODO: Delete crash reports
            // TODO: Delete performance data
            // TODO: Clear session data
            
            // Clear user identification
            sdkState.sessionManager.identify(userId: "", attributes: [:])
            
            Logger.shared.info("User data deleted for user: \(userId)")
            return true
        } catch {
            Logger.shared.error("Error deleting user data", error: error)
            return false
        }
    }
    
    /**
     * Anonymize user data.
     */
    static func anonymizeUserData(userId: String) async -> Bool {
        do {
            // TODO: Anonymize all user events
            // Replace userId with anonymous ID
            let anonymousId = "anonymous_\(Date().timeIntervalSince1970)"
            
            Logger.shared.info("User data anonymized for user: \(userId)")
            return true
        } catch {
            Logger.shared.error("Error anonymizing user data", error: error)
            return false
        }
    }
}

